src/vsr.zig

```diff
@@ -1026,3 +998,88 @@ pub fn quorums(replica_count: u8) struct {
         .view_change = quorum_view_change,
     };
 }
+
+/// The SuperBlock's persisted VSR headers.
+/// One of the following:
+///
+/// - SV headers (consecutive chain)
+/// - DVC headers (disjoint chain)
+pub const ViewChangeHeaders = struct {
+    slice: []const Header,
+
+    pub const BoundedArray = std.BoundedArray(Header, constants.pipeline_prepare_queue_max);
+
+    pub fn init(slice: []const Header) ViewChangeHeaders {
+        ViewChangeHeaders.verify(slice);
+
+        return .{ .slice = slice };
+    }
+
+    pub fn verify(slice: []const Header) void {
+        assert(slice.len > 0);
+        assert(slice.len <= constants.pipeline_prepare_queue_max);
+
+        var child: ?*const Header = null;
+        for (slice) |*header| {
+            assert(header.valid_checksum());
+            assert(header.command == .prepare);
+
+            if (child) |child_header| {
+                assert(header.op < child_header.op);
+                assert(header.view <= child_header.view);
+                assert((header.op + 1 == child_header.op) ==
+                    (header.checksum == child_header.parent));
+                assert(header.timestamp < child_header.timestamp);
+            }
+            child = header;
+        }
+    }
+
+    const ViewRange = struct {
+        min: u32, // inclusive
+        max: u32, // inclusive
+
+        pub fn contains(range: ViewRange, view: u32) bool {
+            return range.min <= view and view <= range.max;
+        }
+    };
+
+    /// Returns the range of possible views (of prepare, not commit) for a message that is part of
+    /// the same log_view as these headers.
```

Comment:
I'm not totally clear on what this function is doing eg I can't parse 'part of the same log_view' - does that just mean 'if the log didn't fork'?.

Reply:
So in the caller context, we have some headers (the superblock's VSR headers) and a `log_view` (the superblock's `log_view`), and those correspond — the VSR headers are either SV or DVC headers corresponding to that `log_view`.

The caller (in `journal.zig`) wants to check for each op in the WAL, whether that op could _possibly_ belong to the superblock's `log_view`. (If it cannot, the op is discarded).

For example, suppose we have this start-view message in our superblock, along with `log_view=c`:
(Numbers indicate ops, letters indicate the view during which the op was prepared)

    3b 4b 5b 6b

- If we have a message `2b` in our WAL, it might be part of `log_view=c`.
- If we have a message `2a` in our WAL, it might be part of `log_view=c`. (`a < b`).
- If we have a message `2c` in our WAL, it cannot possibly be part of `log_view=c`, since `c > b` — `3b` _is_ part of the log, and was prepared before `2c`.
- If we have a message `7b` in our WAL, it cannot possibly be part of `log_view=c`, since if it was, it would have been included in the SV message to begin with.